File: 440-k-th-smallest-in-lexicographical-order/k-th-smallest-in-lexicographical-order.java
class Solution {
    public int findKthNumber(int n, int k) {
        int curr = 1; // Bắt đầu từ 1
        k--; // Vì đã xét curr = 1 là vị trí đầu tiên

        while (k > 0) {
            long steps = countSteps(n, curr, curr + 1);
            if (steps <= k) {
                // Bỏ cả cây con curr → nhảy sang cây tiếp theo
                k -= steps;
                curr++;
            } else {
                // Vẫn nằm trong cây con curr → đi sâu hơn vào tầng dưới
                k--;
                curr *= 10;
            }
        }

        return curr;
    }

    // Đếm số bước từ prefix `n1` đến prefix `n2` (không vượt quá n)
    private long countSteps(int n, long n1, long n2) {
        long steps = 0;
        while (n1 <= n) {
            steps += Math.min(n + 1L, n2) - n1;
            n1 *= 10;
            n2 *= 10;
        }
        return steps;
    }
}

File: 440-k-th-smallest-in-lexicographical-order/k-th-smallest-in-lexicographical-order.go
func findKthNumber(n int, k int) int {
    curr := 1;
	k--; // Đã tính curr = 1 là số đầu tiên rồi

	for k > 0 {
		steps := countSteps(n, curr, curr+1);
		if steps <= k {
			// Bỏ qua cả cây con bắt đầu từ curr
			k -= steps;
			curr++;
		} else {
			// Đi sâu vào cây con curr
			k--;
			curr *= 10;
		}
	}
	return curr;
}

// Đếm số lượng số trong cây con từ prefix curr đến next (không vượt quá n)
func countSteps(n int, curr, next int) int {
	steps := 0;
	for curr <= n {
		steps += min(n+1, next) - curr;
		curr *= 10;
		next *= 10;
	}
	return steps;
}

func min(a, b int) int {
	if a < b {
		return a;
	}
	return b;
}